### ISSUE TITLE
Handle Mongo Disconnection

### ISSUE THREAD

**Issue by badavis** [everyone-counts/mongo-stream#7]

**Comment by g2valdez** [comment]
Mongo-stream has to handle three unique scenarios here:
1) Can't connect on startup.
Proposed solution: retry connection on a set interval (10 seconds?) 
2) Mongo disconnects before a change stream is added.
Proposed solution: retry connection on a set interval (10 seconds?), if requests are sent before connection is established, return an error message to the effect of: "Error: Cant connect to {mongo url}"
3) Mongo disconnects after a change stream is added.
Proposed solution: 
write resume tokens to disk 
cache collection names for all open change streams 
retry connection on a set interval(10 seconds?)
upon successful connections, reopen all cached change streams
if requests are sent before connection is established, return an error message to the effect of: "Error: Cant connect to {mongo url}"

**Comment by badavis** [comment]
1. If the initial connection fails I'm fine with the process exiting. It's likely that mongo isn't reachable or the connection configuration is bad. 

2. Let's go with solution 1. We definitely shouldn't crash in this scenario unless the Mongo Driver hits it's reconnect attempt limit

3. Again let's do solution 1. Cacheing requests to retry never gives us the chance to inform the client that there was a failure if a connection is never re-established.

**Comment by g2valdez** [comment]
Triggering the close event is actually much harder than I anticipated. Disabling WiFi on my machine doesn't trigger it, and doesn't immediately cause mongoDB requests to time out. The only way I've found to reliably trigger the 'close' event is to tear down the database, which, to me, is an unrealistic use case. I added various error catches to mongodb requests, that will stop mongo-stream from continuing execution if an error occurs, but mongo tries really hard to not expose the fact that it lost its connection. I can tell mongo to retry the connection IF (big if) the close event is triggered, but as far as testing it, its a bit of a crap shoot. I think that mongo does a decent job of handling down times of less than 30 seconds, which might be enough for the purposes of this project.
